adapter/pkg/messaging/azure_connection.go

```diff
@@ -37,45 +50,180 @@ func init() {
-	var err error
-	var availableTopics []*servicebus.TopicEntity
+// InitiateBrokerConnectionAndValidate to initiate connection and validate azure service bus constructs to
+// further process
+func InitiateBrokerConnectionAndValidate(eventListeningEndpoint string, componentName string, reconnectRetryCount int,
+	reconnectInterval time.Duration, subscriptionIdleTimeDuration time.Duration) ([]Subscription, error) {
 	logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Trying to connect to azure service bus with " +
 		"connection string " + eventListeningEndpoint)
-	namespace, err := servicebus.NewNamespace(servicebus.NamespaceWithConnectionString(eventListeningEndpoint))
-
-	if err == nil {
+	subscriptionMetaDataList := make([]Subscription, 0)
+	namespace, processError := servicebus.NewNamespace(servicebus.NamespaceWithConnectionString(eventListeningEndpoint))
+	if processError == nil {
+		var topicCreationSuccessful bool
 		logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Successfully received namespace ")
 		topicManager := namespace.NewTopicManager()
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-		availableTopics, err = topicManager.List(ctx)
-		//TODO (dnwick) (add retry logic)
-		if err != nil {
-			logger.LoggerMgw.Errorf("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Error occurred while trying to get topic " +
-				"list from azure service bus :%v", err)
+		var availableTopics []*servicebus.TopicEntity
+		var retryAttemptMessage string
+		for j := 0; j < reconnectRetryCount || reconnectRetryCount == -1; j++ {
+			processError = nil
+			if availableTopics == nil {
+				func() {
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+					availableTopics, processError = topicManager.List(ctx)
+				}()
+				if processError != nil {
+					retryAttemptMessage = ""
+					if reconnectRetryCount > 0 {
+						retryAttemptMessage = "Retry attempt : " + strconv.Itoa(reconnectRetryCount)
+					}
+					logger.LoggerMgw.Errorf("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Error occurred while trying to get topic "+
+						"list from azure service bus :%v." + retryAttemptMessage + " Retrying after %s seconds",
+							processError, retryAttemptMessage, reconnectInterval)
+					time.Sleep(reconnectInterval)
+					continue
+				} else {
+					logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Topic list received ")
+				}
+			}
+			if !topicCreationSuccessful {
+				processError = validateAndCreateTopicForSubscription(availableTopics, namespace)
+			}
+			if processError != nil {
+				retryAttemptMessage = ""
+				if reconnectRetryCount > 0 {
+					retryAttemptMessage = "Retry attempt : " + strconv.Itoa(reconnectRetryCount)
+				}
+				logger.LoggerMgw.Errorf("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] %v " + retryAttemptMessage +
+					" Retrying after %s", processError, reconnectInterval)
+				time.Sleep(reconnectInterval)
+				continue
+			}
```

Comment:
IIUC, this block should be identical to line 75 -> 87 with the operation changed. Shall we extract to a common method?

Reply:
We might need to change the method definition to return status, error

Reply:
Done